File: packages/kits/CorpusKit/Tests/CorpusKitTests/ReindexAbortPathTests.swift
// ReindexAbortPathTests.swift
//
// SS-01 Unit C — reindex abort path, Swift port
//
// Discriminating tests for the abort fix in CorpusContentEngine.reindex:
// when a reindex fails between beginShadowGeneration and publishShadowGeneration,
// abandonShadowGeneration must be called before rethrowing so no vectors are
// left stranded in a generation that will never become visible.
//
// Test inventory:
//
//   c4a. Error propagates: the caller receives the original error from the
//        failed reindex, not an error from the abort call.
//
//   c4b. No leftover 'building' state: after a failed reindex, the
//        vector_generations registry must NOT show shadow_state = 'building'.
//        Pre-fix, shadow_state is 'building'; this assertion fails.
//        Post-fix, abandonShadowGeneration clears shadow_state to NULL.
//
//   c4c. Serving generation intact: serving_generation remains at its prior
//        value (the old generation keeps serving), same as the c3 check in
//        ShadowSwapCorpusTests but retained here alongside the c4b assertion.
//
//   c4d. No stranded shadow vectors: the vector count for the model equals
//        the count before the failed reindex. Failure during trainTrainableSlots
//        (before the write pass) means no shadow vectors were written, so the
//        count stays at 6 (3 items × 2 lanes). Abandon would delete any that
//        were written; the count check confirms no third-state rows remain.
//
// Seam used: a CorpusContentSource that throws on the Nth call to record(for:).
// No production code is added purely to make this testable; the source
// interface already exposes the injection point.
//
// Storage: on-disk SQLite via makeScratchStorage() — the same backend
// production uses.

import Testing
import Foundation
import SynapseKit
import PersistenceKit
import CorpusKitProviders
@testable import CorpusKit

// MARK: - Helpers (file-private; not reused from ShadowSwapCorpusTests)

/// In-memory content source that can be armed to throw on the next record(for:) call.
private actor FaultingContentSource: CorpusContentSource {
    private var records: [String: CorpusContentRecord] = [:]
    private var throwOnNextRecord = false

    func put(id: String, text: String) {
        let digest = CorpusContentDigest.digest(text)
        let revision: Int64 = (records[id].map { $0.revision } ?? 0) + 1
        records[id] = CorpusContentRecord(
            id: id, revision: revision, digest: digest, text: text)
    }

    /// Arm the source: the NEXT call to record(for:) throws and disarms.
    func armFault() {
        throwOnNextRecord = true
    }

    func record(for id: CorpusContentID) async throws -> CorpusContentRecord? {
        if throwOnNextRecord {
            throwOnNextRecord = false
            throw FaultError.forcedRecordFailure
        }
        return records[id]
    }

    func activeContentIDs() async throws -> [CorpusContentID] {
        records.keys.sorted()
    }

    func changes(since cursor: String?, limit: Int) async throws -> CorpusContentChangeBatch {
        // Reindex tests go through activeContentIDs + record(for:), not the feed.
        .empty
    }

    private enum FaultError: Error {
        case forcedRecordFailure
    }
}

/// Read the `shadow_state` column for a model from `vector_generations`.
/// Returns nil when no row exists or when shadow_state is SQL NULL.
private func shadowState(storage: any Storage, modelID: String) async throws -> String? {
    let rows = try await storage.rowStore.query(
        table: "vector_generations",
        where: .eq(Column(table: "vector_generations", name: "model_id"), .text(modelID)),
        orderBy: [],
        limit: 1,
        offset: nil
    )
    guard let row = rows.first else { return nil }
    switch row.values["shadow_state"] {
    case .text(let s): return s
    default: return nil  // NULL or absent: shadow_state has been cleared
    }
}

/// Read `serving_generation` for a model. Returns nil when no registry row exists.
private func servingGeneration(storage: any Storage, modelID: String) async throws -> Int64? {
    let rows = try await storage.rowStore.query(
        table: "vector_generations",
        where: .eq(Column(table: "vector_generations", name: "model_id"), .text(modelID)),
        orderBy: [],
        limit: 1,
        offset: nil
    )
    guard let row = rows.first else { return nil }
    switch row.values["serving_generation"] {
    case .int(let v): return v
    case .text(let s): return Int64(s)
    default: return nil
    }
}

/// Count ALL rows in `vectors` for a model, regardless of generation.
private func vectorRowCount(storage: any Storage, modelID: String) async throws -> Int {
    let rows = try await storage.rowStore.query(
        table: "vectors",
        where: .eq(Column(table: "vectors", name: "model_id"), .text(modelID)),
        orderBy: [],
        limit: nil,
        offset: nil
    )
    return rows.count
}

/// Build a CorpusContentEngine with one trainable RI slot over a given storage
/// and source. The RI slot causes beginShadowGeneration to engage on every
/// reindex, so the shadow-abort path is exercised.
private func makeAbortTestEngine(
    storage: any Storage,
    source: any CorpusContentSource
) async throws -> CorpusContentEngine {
    let config = try CorpusContentConfiguration(mode: .standalone, indexUnit: .wholeContent)
    let provider = RandomIndexingProvider()
    return try await CorpusContentEngine(
        storage: storage,
        configuration: config,
        source: source,
        models: [.randomIndexing(provider: provider)]
    )
}

// MARK: - Suite

@Suite("SS-01 reindex abort path — abandonShadowGeneration called on failure",
       .serialized)
struct ReindexAbortPathTests {

    // ── c4a + c4b + c4c + c4d: all four assertions in one scenario ────────────

    /// Scenario: first reindex succeeds (serving_gen=1, 6 vectors). Then the
    /// source is armed to throw on the NEXT record(for:) call, so the second
    /// reindex fails inside trainTrainableSlots (after beginShadowGeneration,
    /// before any shadow vectors are written, and well before publish).
    ///
    /// Expected state after the failed second reindex:
    ///
    ///   c4a: the error propagates — the caller sees an error, not nil.
    ///
    ///   c4b: shadow_state is NOT 'building'. Pre-fix, the registry row is
    ///        left with shadow_state='building', so this assertion FAILS before
    ///        the C1 fix is applied. Post-fix, abandonShadowGeneration sets
    ///        shadow_state to NULL.
    ///
    ///   c4c: serving_generation is still 1 — the old generation keeps serving.
    ///
    ///   c4d: vector count is still 6 — no shadow vectors were stranded (the
    ///        failure occurred before the write pass, and abandon would have
    ///        cleaned up any that were written).
    ///
    /// This test FAILS on the current code (pre-fix c4b), which demonstrates
    /// the bug: `shadow_state='building'` is the orphaned third state the
    /// mission's governing invariant forbids.
    @Test("c4 failed reindex: error propagates and no leftover building shadow in registry")
    func c4_failedReindexAbandonsOrphanedShadow() async throws {
        let modelID = "random-indexing-v1"
        let storage = try makeScratchStorage()
        let source = FaultingContentSource()
        await source.put(id: "doc-1", text: "the quick brown fox jumps over the lazy dog")
        await source.put(id: "doc-2", text: "machine learning and natural language processing")
        await source.put(id: "doc-3", text: "cats and dogs are common household pets")

        let engine = try await makeAbortTestEngine(storage: storage, source: source)
        let now = Date(timeIntervalSinceReferenceDate: 25_000_000)

        // ── First reindex: success — establishes serving_gen=1, 6 vectors ────
        try await engine.reindex(now: now)

        let gen1 = try await servingGeneration(storage: storage, modelID: modelID)
        #expect(gen1 == 1, "setup: serving_generation must be 1 after first reindex")
        let countAfterFirst = try await vectorRowCount(storage: storage, modelID: modelID)
        #expect(countAfterFirst == 6, "setup: 6 vectors after first reindex (3 items × 2 lanes)")

        // ── Arm fault: next record(for:) call throws ──────────────────────────
        // This fires inside trainTrainableSlots on the second reindex, which runs
        // after beginShadowGeneration. The shadow is open (shadow_state='building')
        // when the error propagates.
        await source.armFault()

        // ── Second reindex: must propagate the error (c4a) ───────────────────
        // Pre-fix: leaves shadow_state='building' in the registry.
        // Post-fix: abandonShadowGeneration is called, shadow_state cleared.
        await #expect(throws: (any Error).self) {
            try await engine.reindex(now: now)
        }

        // ── c4b: shadow_state must NOT be 'building' after the abort ─────────
        // Before the fix, shadow_state='building' — this assertion FAILS, which
        // demonstrates the bug. After the fix, shadow_state is NULL.
        let state = try await shadowState(storage: storage, modelID: modelID)
        #expect(
            state != "building",
            """
            c4b: shadow_state must not be 'building' after a failed reindex — \
            pre-fix this was 'building' (the orphaned third state). \
            Got: \(state ?? "nil")
            """
        )

        // ── c4c: serving_generation unchanged — old generation still serves ───
        let genAfterFail = try await servingGeneration(storage: storage, modelID: modelID)
        #expect(
            genAfterFail == 1,
            "c4c: serving_generation must remain 1 when second reindex fails before publish"
        )

        // ── c4d: vector count unchanged — no vectors stranded at shadow gen ───
        // Failure was before the write pass, so no shadow vectors were written.
        // Abandon confirms there is nothing to clean up; count stays at 6.
        let countAfterFail = try await vectorRowCount(storage: storage, modelID: modelID)
        #expect(
            countAfterFail == 6,
            "c4d: vector count must remain 6 — no shadow vectors stranded at an invisible generation"
        )
    }
}

File: packages/kits/CorpusKit/rust/tests/reindex_abort_path_tests.rs
//! reindex_abort_path_tests.rs
//!
//! SS-01 Unit C — reindex abort path, Rust port.
//!
//! Discriminating tests for the abort fix in ContentEngine::reindex: when a
//! reindex fails between begin_shadow_generation and publish_shadow_generation,
//! abandon_shadow_generation must be called before propagating the error so no
//! vectors are stranded in a generation that will never become visible.
//!
//! Test inventory (mirrors ReindexAbortPathTests.swift):
//!
//!   c4a. Error propagates — the caller receives the original Err, not Ok.
//!
//!   c4b. No leftover 'building' state — after a failed reindex, the
//!        vector_generations registry must NOT show shadow_state = 'building'.
//!        Pre-fix, shadow_state is 'building'; this assertion fails.
//!        Post-fix, abandon_shadow_generation clears shadow_state to NULL.
//!
//!   c4c. Serving generation intact — serving_generation remains at its prior
//!        value; the old generation keeps serving.
//!
//!   c4d. No stranded shadow vectors — vector count for the model is unchanged.
//!        Failure during train_trainable_slots (before the write pass) means no
//!        shadow vectors were written; count stays at 6 (3 items × 2 lanes).
//!
//! Seam used: a CorpusContentSource implementation that returns Err on the Nth
//! call to record(). No production code is modified to make this testable.
//!
//! Storage: on-disk SQLite (RAII temp dir) — the same backend production uses.

use corpus_kit::{
    CorpusContentChangeBatch, CorpusContentConfiguration, CorpusContentEngine,
    CorpusContentId, CorpusContentRecord, CorpusContentSource, CorpusIndexUnitPolicy,
    CorpusKitError, CorpusOperatingMode, EmbeddingModelConfig,
};
use corpus_kit::content_digest;
use corpus_kit_providers::RandomIndexingProvider;
use persistence_kit::{
    BackendConfiguration, Column, EstateConfiguration, SqliteStorage, Storage,
    StoragePredicate, TypedValue,
};
use std::collections::BTreeMap;
use std::path::PathBuf;
use std::sync::{Arc, Mutex};
use std::sync::atomic::{AtomicBool, Ordering};
use uuid::Uuid;

// ── Constants ──────────────────────────────────────────────────────────────

const NOW_MILLIS: i64 = 1_755_100_000_000;
const RI_MODEL_ID: &str = "random-indexing-v1";

// ── RAII temp dir ──────────────────────────────────────────────────────────

struct TempDir(PathBuf);
impl Drop for TempDir {
    fn drop(&mut self) {
        let _ = std::fs::remove_dir_all(&self.0);
    }
}

fn make_scratch_storage() -> (Arc<dyn Storage>, TempDir) {
    let dir = std::env::temp_dir().join(format!("reindex-abort-rs-{}", Uuid::new_v4()));
    std::fs::create_dir_all(&dir).expect("create temp dir");
    let path = dir
        .join(format!("reindex-abort-{}.sqlite3", Uuid::new_v4()))
        .to_string_lossy()
        .into_owned();
    let config = EstateConfiguration::new(
        Uuid::new_v4(),
        BackendConfiguration::Sqlite { path, busy_timeout_secs: 5.0 },
    );
    let storage: Arc<dyn Storage> =
        Arc::new(SqliteStorage::new(config).expect("open SQLite storage"));
    (storage, TempDir(dir))
}

// ── FaultSource ─────────────────────────────────────────────────────────────

/// CorpusContentSource that can be armed to fail on the next record() call.
struct FaultSource {
    records: Mutex<BTreeMap<String, CorpusContentRecord>>,
    /// When true, the next record() call returns Err and resets the flag.
    fail_on_next_record: AtomicBool,
}

impl FaultSource {
    fn new() -> Arc<Self> {
        Arc::new(Self {
            records: Mutex::new(BTreeMap::new()),
            fail_on_next_record: AtomicBool::new(false),
        })
    }

    fn put(&self, id: &str, text: &str) {
        let mut r = self.records.lock().unwrap();
        r.insert(
            id.to_string(),
            CorpusContentRecord {
                id: id.to_string(),
                revision: 1,
                digest: content_digest(text),
                text: text.to_string(),
                dense_composition_text: None,
            },
        );
    }

    /// Arm: the next record() call will return Err (single-use).
    fn arm_fault(&self) {
        self.fail_on_next_record.store(true, Ordering::SeqCst);
    }
}

impl CorpusContentSource for FaultSource {
    fn record(&self, id: &str) -> Result<Option<CorpusContentRecord>, CorpusKitError> {
        if self.fail_on_next_record.swap(false, Ordering::SeqCst) {
            return Err(CorpusKitError::StoreUnavailable(
                "FaultSource: forced record() failure for abort path test".into(),
            ));
        }
        Ok(self.records.lock().unwrap().get(id).cloned())
    }

    fn changes(
        &self,
        _cursor: Option<&str>,
        _limit: usize,
    ) -> Result<CorpusContentChangeBatch, CorpusKitError> {
        Ok(CorpusContentChangeBatch::empty())
    }

    fn active_content_ids(&self) -> Result<Vec<CorpusContentId>, CorpusKitError> {
        let r = self.records.lock().unwrap();
        let mut ids: Vec<String> = r.keys().cloned().collect();
        ids.sort();
        Ok(ids)
    }
}

// ── Storage inspection helpers ──────────────────────────────────────────────

/// Read shadow_state for model_id from vector_generations. Returns None when
/// no row exists or when shadow_state is SQL NULL.
fn shadow_state(storage: &dyn Storage, model_id: &str) -> Option<String> {
    let rows = storage
        .row_store()
        .query(
            "vector_generations",
            Some(&StoragePredicate::Eq(
                Column::new("vector_generations", "model_id"),
                TypedValue::Text(model_id.to_string()),
            )),
            &[],
            Some(1),
            None,
        )
        .expect("query vector_generations");
    let row = rows.first()?;
    match row.get("shadow_state") {
        Some(TypedValue::Text(s)) => Some(s.clone()),
        _ => None,  // NULL or absent: shadow_state has been cleared
    }
}

/// Read serving_generation for model_id. Returns None when no registry row exists.
fn serving_generation(storage: &dyn Storage, model_id: &str) -> Option<i64> {
    let rows = storage
        .row_store()
        .query(
            "vector_generations",
            Some(&StoragePredicate::Eq(
                Column::new("vector_generations", "model_id"),
                TypedValue::Text(model_id.to_string()),
            )),
            &[],
            Some(1),
            None,
        )
        .expect("query vector_generations");
    let row = rows.first()?;
    match row.get("serving_generation") {
        Some(TypedValue::Int(v)) => Some(*v),
        Some(TypedValue::Text(s)) => s.parse().ok(),
        _ => None,
    }
}

/// Count ALL rows in vectors for model_id, regardless of generation.
fn vector_row_count(storage: &dyn Storage, model_id: &str) -> usize {
    storage
        .row_store()
        .query(
            "vectors",
            Some(&StoragePredicate::Eq(
                Column::new("vectors", "model_id"),
                TypedValue::Text(model_id.to_string()),
            )),
            &[],
            None,
            None,
        )
        .expect("query vectors")
        .len()
}

// ── Engine factory ──────────────────────────────────────────────────────────

fn make_abort_test_engine(
    storage: Arc<dyn Storage>,
    source: Arc<FaultSource>,
) -> CorpusContentEngine {
    let config = CorpusContentConfiguration::new(
        CorpusOperatingMode::Standalone,
        CorpusIndexUnitPolicy::WholeContent,
    )
    .expect("abort test engine configuration");

    CorpusContentEngine::open(
        storage,
        config,
        source as Arc<dyn CorpusContentSource>,
        vec![EmbeddingModelConfig::RandomIndexing {
            provider: Box::new(RandomIndexingProvider::new()),
        }],
        false,
    )
    .expect("abort test engine open")
}

// ── c4: all four assertions ──────────────────────────────────────────────────

/// Scenario: first reindex succeeds (serving_gen=1, 6 vectors). Then the
/// source is armed to throw on the next record() call, so the second reindex
/// fails inside train_trainable_slots (after begin_shadow_generation, before
/// any shadow vectors are written, and before publish_shadow_generation).
///
/// Expected state after the failed second reindex:
///
///   c4a: the error propagates — reindex returns Err, not Ok.
///
///   c4b: shadow_state is NOT 'building'. Pre-fix, the registry row is left
///        with shadow_state='building', so this assertion FAILS before the
///        C2 fix is applied. Post-fix, abandon_shadow_generation sets
///        shadow_state to NULL.
///
///   c4c: serving_generation remains 1 — old generation still serves.
///
///   c4d: vector count remains 6 — no vectors stranded at shadow generation.
///
/// This test FAILS on the current code (pre-fix c4b), demonstrating the bug:
/// shadow_state='building' is the orphaned third state the governing invariant
/// forbids.
#[test]
fn c4_failed_reindex_abandons_orphaned_shadow() {
    let (storage, _guard) = make_scratch_storage();
    let source = FaultSource::new();
    source.put("doc-1", "the quick brown fox jumps over the lazy dog");
    source.put("doc-2", "machine learning and natural language processing");
    source.put("doc-3", "cats and dogs are common household pets");

    let engine = make_abort_test_engine(Arc::clone(&storage), Arc::clone(&source));

    // ── First reindex: success — establishes serving_gen=1, 6 vectors ────────
    engine.reindex(NOW_MILLIS).expect("first reindex");

    let gen1 = serving_generation(&*storage, RI_MODEL_ID);
    assert_eq!(gen1, Some(1), "setup: serving_generation must be 1 after first reindex");

    let count_after_first = vector_row_count(&*storage, RI_MODEL_ID);
    assert_eq!(count_after_first, 6, "setup: 6 vectors after first reindex (3 items × 2 lanes)");

    // ── Arm fault: next record() call throws ─────────────────────────────────
    // This fires inside train_trainable_slots on the second reindex, which runs
    // after begin_shadow_generation. The shadow is open (shadow_state='building')
    // when the error propagates.
    source.arm_fault();

    // ── Second reindex: must return Err (c4a) ────────────────────────────────
    // Pre-fix: leaves shadow_state='building' in the registry.
    // Post-fix: abandon_shadow_generation is called, shadow_state cleared.
    let result = engine.reindex(NOW_MILLIS);
    assert!(
        result.is_err(),
        "c4a: second reindex must return Err when the source fails"
    );

    // ── c4b: shadow_state must NOT be 'building' ─────────────────────────────
    // Before the fix, shadow_state='building' and this assertion FAILS, which
    // demonstrates the bug. After the fix, shadow_state is NULL (None here).
    let state = shadow_state(&*storage, RI_MODEL_ID);
    assert_ne!(
        state.as_deref(),
        Some("building"),
        "c4b: shadow_state must not be 'building' after a failed reindex — \
         pre-fix this was 'building' (orphaned third state). Got: {:?}",
        state
    );

    // ── c4c: serving_generation unchanged — old generation still serves ───────
    let gen_after_fail = serving_generation(&*storage, RI_MODEL_ID);
    assert_eq!(
        gen_after_fail,
        Some(1),
        "c4c: serving_generation must remain 1 when second reindex fails before publish"
    );

    // ── c4d: vector count unchanged — no vectors stranded at shadow gen ────────
    let count_after_fail = vector_row_count(&*storage, RI_MODEL_ID);
    assert_eq!(
        count_after_fail, 6,
        "c4d: vector count must remain 6 — no shadow vectors stranded at invisible generation"
    );
}
